lib/awful/client.lua

```diff
@@ -439,72 +439,90 @@ function client.object.move_to_screen(self, s)
+--
+-- @beautiful beautiful.border_marked
+-- @param string
+--
+
+function client.object.set_marked(self, value)
+    local is_marked = self.marked
+
+    if value == false and is_marked then
+        for k, v in pairs(client.data.marked) do
+            if self == v then
+                table.remove(client.data.marked, k)
             end
         end
+        self:emit_signal("unmarked")
+    elseif not is_marked and value then
+        self:emit_signal("marked")
+        table.insert(client.data.marked, self)
+    end
 
-        table.insert(client.data.marked, cl)
+    client.property.set(self, "marked", value)
+end
 
-        -- Call callback
-        cl:emit_signal("marked")
-        return true
-    end
+function client.object.get_marked(self)
+    return client.property.get(self, "marked")
 end
 
+--- Mark a client, and then call 'marked' hook.
+-- @deprecated awful.client.mark
+-- @client c The client to mark, the focused one if not specified.
+function client.mark(c)
+    util.deprecate "Use c.marked = true instead of awful.client.mark"
 
+    client.object.set_marked(c or capi.client.focus, true)
+end
 
 --- Unmark a client and then call 'unmarked' hook.
+-- @deprecated awful.client.unmark
 -- @client c The client to unmark, or the focused one if not specified.
--- @return True if the client has been unmarked. False if the client was not marked.
 function client.unmark(c)
-    local cl = c or capi.client.focus
-
-    for k, v in pairs(client.data.marked) do
-        if cl == v then
-            table.remove(client.data.marked, k)
-            cl:emit_signal("unmarked")
-            return true
-        end
-    end
+    util.deprecate "Use c.marked = false instead of awful.client.unmark"
 
-    return false
+    client.object.set_marked(c or capi.client.focus, false)
 end
 
 --- Check if a client is marked.
+-- @deprecated awful.client.ismarked
 -- @client c The client to check, or the focused one otherwise.
 function client.ismarked(c)
-    local cl = c or capi.client.focus
-    if cl then
-        for _, v in pairs(client.data.marked) do
-            if cl == v then
-                return true
-            end
-        end
-    end
-    return false
+    util.deprecate "Use c.marked instead of awful.client.ismarked"
+
+    return client.object.get_marked(c or capi.client.focus)
 end
 
 --- Toggle a client as marked.
+-- @deprecated awful.client.togglemarked
 -- @client c The client to toggle mark.
 function client.togglemarked(c)
+    util.deprecate "Use c.marked = not c.marked instead of awful.client.togglemarked"
+
     c = c or capi.client.focus
-    if not client.mark(c) then
-        client.unmark(c)
+    if c then
+        c.marked = not c.marked
     end
 end
 
 --- Return the marked clients and empty the marked table.
+-- @function awful.client.getmarked
 -- @return A table with all marked clients.
 function client.getmarked()
     for _, v in pairs(client.data.marked) do
+        client.property.set(v, "marked", false)
```

Comment:
I don't think this works. `pairs()` remembers the key that it returned last (1), then you remove that entry. Since this is in the array-part of the table, everything is shifted down. So what was previously entry 2 is now entry 1. Now, `pairs` looks at the next entry after 1 and the old-entry-2-now-entry-1 is missed.
(I haven't actually tested this, it's just a theory)

Also, the return value of this function will be wrong since by the time it gets `client.data.marked`, that table will already be empty.